src/Integration.UnitTests/Binding/SolutionBindingOperationTests.cs

```diff
@@ -268,85 +284,51 @@ public void SolutionBindingOperation_Prepare()
         {
             // Arrange
-            var csProject = this.solutionMock.AddOrGetProject("CS.csproj");
-            csProject.SetCSProjectKind();
+            var csProject = solutionMock.AddOrGetProject("CS.csproj");
+            var csConfigFile = CreateMockConfigFile("c:\\csharp.txt");
+            var csBinder = new Mock<IProjectBinder>();
+
             var vbProject = this.solutionMock.AddOrGetProject("VB.vbproj");
-            vbProject.SetVBProjectKind();
+            var vbConfigFile = CreateMockConfigFile("c:\\vb.txt");
+            var vbBinder = new Mock<IProjectBinder>();
+
+            projectBinderFactoryMock.Setup(x => x.Get(csProject)).Returns(csBinder.Object);
+            projectBinderFactoryMock.Setup(x => x.Get(vbProject)).Returns(vbBinder.Object);
+
             var projects = new[] { csProject, vbProject };
 
-            SolutionBindingOperation testSubject = this.CreateTestSubject(ProjectKey);
+            var testSubject = CreateTestSubject(ProjectKey);
 
-            var csConfigFile = CreateMockRuleSetConfigFile(Language.CSharp, "c:\\csharp.txt");
-            var vbConfigFile = CreateMockRuleSetConfigFile(Language.VBNET, "c:\\vb.txt");
             var languageToFileMap = new Dictionary<Language, IBindingConfigFile>();
             languageToFileMap[Language.CSharp] = csConfigFile.Object;
             languageToFileMap[Language.VBNET] = vbConfigFile.Object;
 
             testSubject.RegisterKnownConfigFiles(languageToFileMap);
             testSubject.Initialize(projects, GetQualityProfiles());
-            testSubject.Binders.Clear(); // Ignore the real binders, not part of this test scope
-            bool prepareCalledForBinder = false;
+
             using (CancellationTokenSource src = new CancellationTokenSource())
             {
-                testSubject.Binders.Add(new ConfigurableBindingOperation { PrepareAction = (t) => src.Cancel() });
-                testSubject.Binders.Add(new ConfigurableBindingOperation { PrepareAction = (t) => prepareCalledForBinder = true });
-
+                src.Cancel();
                 // Act
                 testSubject.Prepare(src.Token);
             }
 
             // Assert
+            testSubject.Binders.Count.Should().Be(0);
+            csBinder.VerifyNoOtherCalls();
+            vbBinder.VerifyNoOtherCalls();
+
             testSubject.RuleSetsInformationMap[Language.CSharp].FilePath.Should().Be("c:\\csharp.txt");
             testSubject.RuleSetsInformationMap[Language.VBNET].FilePath.Should().Be("c:\\vb.txt");
-            prepareCalledForBinder.Should().BeFalse("Expected to be canceled as soon as possible i.e. after the first binder");
-
-            CheckSaveWasNotCalled(csConfigFile);
-            CheckSaveWasNotCalled(vbConfigFile);
-        }
-
-        [TestMethod]
-        public void SolutionBindingOperation_Prepare_Cancellation_BeforeBindersPrepare()
```

Comment:
It was difficult to test the two use-cases (beforeBindersPrepare and duringBindersPrepare), so I left only beforeBindersPrepare.

Reply:
Couldn't you provide a mocked implementation of `projectBinder.GetBindAction` that cancels the operation? i.e. provide two binders, the first of which cancels the operation, then check the second wasn't called.

Or perhaps simpler - cancel the operation in  `sourceControlledFileSystem.QueueFileWrite`, then check the binders aren't called.